File: conn/odbc/src/odbc/Common/compression.cpp
// @@@ START COPYRIGHT @@@
//
// (C) Copyright 2003-2014 Hewlett-Packard Development Company, L.P.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//
// @@@ END COPYRIGHT @@@

#include "compression.h"
#include <zlib.h>

CCompression::CCompression()
{
}

CCompression::~CCompression()
{
}

bool CCompression::compress(unsigned char *input, unsigned long input_size, int level, unsigned char **output, unsigned long& output_size)
{
	int ret, flush;
	z_stream strm;
	strm.zalloc = Z_NULL;
	strm.zfree = Z_NULL;
	strm.opaque = Z_NULL;
	ret = deflateInit(&strm, level);
	if (ret != Z_OK)
		return ret;

	strm.next_in = input;
	strm.avail_in = input_size;
	strm.data_type = Z_BINARY;

	output_size = deflateBound(&strm,input_size);
	*output = new unsigned char[output_size];
	if(!*output)
	{
		deflateEnd(&strm);
		return false;
	}

	strm.avail_out = output_size;
	strm.next_out = (Bytef*)*output;
	flush = Z_FINISH;

	ret = deflate(&strm, flush);
	deflateEnd(&strm);
	if(ret == Z_STREAM_END)  /* have to be Z_STREAM_END */
	{
		if((output_size-strm.avail_out)>=input_size)
		{
			return false;
		}
		output_size-=strm.avail_out;
		return true;
	}
	else
		return false;
}

bool CCompression::expand(unsigned char *input, unsigned long input_size, unsigned char **output, unsigned long& output_size, int& error)
{
	int ret;
    z_stream strm;

	if(!input_size)
	{
		output_size = 0;
		return false;
	}

	strm.zalloc = Z_NULL;
    strm.zfree = Z_NULL;
    strm.opaque = Z_NULL;
    strm.avail_in = 0;
    strm.next_in = Z_NULL;
    ret = inflateInit(&strm);
	if (ret != Z_OK)
        return ret;

	strm.next_in = input;
	strm.avail_in = input_size;
	strm.next_out = (Bytef*)*output;
	strm.avail_out=output_size;		 
    ret = inflate(&strm, Z_FINISH);

	inflateEnd(&strm);

	if(ret == Z_STREAM_END && strm.avail_out == 0)
	{
		error=0;
		return true;
	}
   	
    error=ret;
	output_size = 0;
	return false;
}

File: core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/Compression.java
// @@@ START COPYRIGHT @@@
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
// @@@ END COPYRIGHT @@@
package org.trafodion.jdbc.t4;

public class Compression {
	private static int MAXCOUNT = 0xFF;
	// private static int MAXUCOUNT = 0xFFFF;
	private static int EOFCHAR = -1;
	
	
	
	public static int compress(int alg, byte [][] sbuf, int [] sstart, int [] slen, byte[] dbuf, int dstart) {	
		int ret = 0;
		
		switch(alg) {
		case 1:
			ret = compress_whitespace(sbuf, sstart, slen, dbuf, dstart);
			break;
		}
		
		return ret;
	}
	
	public static int uncompress(int alg, byte [] sbuf, int sstart, int slen, byte [] dbuf, int dstart) {
		int ret = 0;
		
		switch(alg) {
		case 1:
			ret = uncompress_whitespace(sbuf, sstart, slen, dbuf, dstart);
			break;
		}
		
		return ret;
	}
	
	// encodes repeated bytes in the byte form <b><b><#>
	// b = byte which is repeated
	// # = number of repetitions (max 255)
	// source buffers, source lengths, destination buffer
	private static int compress_whitespace(byte [][] param_sbuf, int [] param_sstart, int [] param_slen, byte[] dbuf, int dstart) {
		int c = EOFCHAR, p = EOFCHAR; 
		int si = 0, di = dstart; // source, destination indexes
		int cnt = 0; // count of repetition
		
		byte []sbuf; //current source buffer
		int slen; //current source length
		
		for(int i=0;i<param_sbuf.length;i++) {
			sbuf = param_sbuf[i];
			slen = param_slen[i];
			si = param_sstart[i];
			
			while(si < slen) {
				c = sbuf[si++]; // get the next byte from source
				dbuf[di++] = (byte)c; // copy the byte to destination
				cnt = 0; 
				
				if(c == p) { // check repetition
					if(si == slen) {
						c = EOFCHAR;
					}
					while(si < slen) {
		                if ((c = sbuf[si++]) == p) { // found repetition
		                    cnt++;
		                    if (cnt == MAXCOUNT) { // we can only store 255 in
													// a byte
		                        dbuf[di++] = (byte)cnt;
		                        
		                        p = EOFCHAR; // move on
		                        break;
		                    }
		                }
		                else {
		                    dbuf[di++] = (byte)cnt;
		                    dbuf[di++] = (byte) c;
	
		                    p = c;
		                    break;
		                }
		                
		                if (si == slen) {
	    	            	c = EOFCHAR;
	    	            }
		            } 
				}
				else {
					p = c; // set our current as our previous
				}
				
				if (c == EOFCHAR) {
 	            	dbuf[di++] = (byte)cnt;
 	            	/*
 	            	 * ADDED
 	            	 */
 	            	p = EOFCHAR;
 	            }
			}
		}
		
		return di - dstart;
	}
	
	private static int uncompress_whitespace(byte [] sbuf, int sstart, int slen, byte [] dbuf, int dstart) {
		int c = EOFCHAR, p = EOFCHAR;
		int si = sstart, di = dstart, i = 0; // source, dest, and generic indexes
		int cnt = 0;

		while(si < slen ) {
			c = sbuf[si++] & 0xFF;
			
			/*if(di >= dbuf.length) {
				System.out.println("%%%%%failed at " + di);
			}*/
			dbuf[di++] = (byte) c;
			
			if(c == p) {
				cnt = sbuf[si++] & 0xFF;
				
				for(i=0;i<cnt;++i) {
					dbuf[di++] = (byte)c;
				}
				p = EOFCHAR;
			}
			else {
				p = c;
			}
		}
		
		return di;
	}	
}
